Upgrade samples

jngrad opened the pull request:
Fixes #2089

Description of changes:
- remove unused variables/imports/functions and some of the commented out code
- remove Tcl-specific code, replace manual warmup by steepest descent
- use `argparse` to process user input and document the parameters, stop printing features
- improve docstrings and use them to generate an always up-to-date sample list in the docs

jngrad committed:
Use a class

jngrad committed:
Formatting

jngrad committed:
Document samples

jngrad committed:
Cleanup samples

Remove unused imports/variables/functions, fix mistakes, simplify
expressions, replace some magic numbers by expressions, replace
sys.stdout() by print(), remove commented code.

jngrad committed:
Use argparse

jngrad committed:
Don't print list of compiled features

jngrad committed:
Generate samples documentation from docstrings

KaiSzuttor commented on samples/lj-demo.py:
```diff
@@ -132,7 +136,7 @@
 
 act_min_dist = system.analysis.min_dist()
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
where does this magic number come from?

jngrad replied:
I've removed all occurrences of `max_num_cells`. There was no performance impact.

KaiSzuttor commented on samples/lj_liquid.py:
```diff
@@ -45,7 +48,7 @@
 box_l = 10.7437
```
please remove the outdated comment above this line

KaiSzuttor commented on samples/lj_liquid.py:
```diff
@@ -93,21 +96,21 @@
 print("Start with minimal distance {}".format(act_min_dist))
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
magic number?

KaiSzuttor commented on samples/lj_liquid.py:
```diff
@@ -133,22 +136,16 @@
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
maybe we should ```minimize_energy``` here

RudolfWeeber replied:
If we should, we should system.integrator.set_steepest_descent()

jngrad replied:
Done in 23d39c7 for all samples.

KaiSzuttor commented on samples/lj_liquid_distribution.py:
```diff
@@ -48,7 +45,7 @@
 box_l = 10.7437
```
please remove the outdated comment above this line

KaiSzuttor commented on samples/lj_liquid_distribution.py:
```diff
@@ -125,13 +120,13 @@
 print("Start with minimal distance {}".format(act_min_dist))
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
magic number?

KaiSzuttor commented on samples/lj_liquid_structurefactor.py:
```diff
@@ -114,13 +118,13 @@
 print("Start with minimal distance {}".format(act_min_dist))
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
magic number?

KaiSzuttor commented on samples/lj_liquid_structurefactor.py:
```diff
@@ -143,18 +147,16 @@
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
```minimize_energy```?

KaiSzuttor commented on samples/minimal-charged-particles.py:
```diff
@@ -92,7 +92,7 @@
 i = 0
 act_min_dist = system.analysis.min_dist()
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
```minimize_energy```?

KaiSzuttor commented on samples/minimal-diamond.py:
```diff
@@ -90,15 +84,14 @@
 
 # slowly ramp up the cap
-while (wca_cap < 5):
+while wca_cap < 5:
```
```minimize_energy```?

KaiSzuttor commented on samples/minimal-polymer.py:
```diff
@@ -67,26 +65,24 @@
 
 # slowly ramp un up the cap
-while (act_min_dist < 0.95):
+while act_min_dist < 0.95:
```
```minimize_energy```?

KaiSzuttor commented on samples/p3m.py:
```diff
@@ -159,7 +175,7 @@
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times or act_min_dist < min_dist):
+while i < warm_n_times or act_min_dist < min_dist:
```
```minimize_energy```?

KaiSzuttor commented on samples/reaction_ensemble.py:
```diff
@@ -53,7 +54,7 @@
 system.time_step = 0.02
 system.cell_system.skin = 0.4
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
magic number?

KaiSzuttor commented on samples/rigid_body.py:
```diff
@@ -16,21 +16,19 @@
 box_l = 100
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.set_random_state_PRNG()
```
this is almost never needed (unless you use NPT)

KaiSzuttor commented on samples/slice_input.py:
```diff
@@ -47,7 +44,7 @@
 system.cell_system.skin = 0.4
 
-system.cell_system.max_num_cells = 2744
+system.cell_system.max_num_cells = 14**3
```
magic number?

KaiSzuttor commented on samples/visualization_ljliquid.py:
```diff
@@ -136,36 +137,16 @@
 # Warmup Integration Loop
 i = 0
-while (i < warm_n_times and act_min_dist < min_dist):
+while i < warm_n_times and act_min_dist < min_dist:
```
```minimize_energy```?

KaiSzuttor commented on samples/visualization_ljliquid.py:
```diff
@@ -186,12 +167,10 @@
 def main_loop():
     global energies
-    print("run %d at time=%f " % (i, system.time))
+    print("run at time=%f " % system.time)
```
we should probably also use ```"{}".format(```

KaiSzuttor commented on samples/visualization_poiseuille.py:
```diff
@@ -15,20 +15,20 @@
 box_l = 16
-system = System(box_l=[box_l] * 3)
+system = espressomd.System(box_l=[box_l] * 3)
 system.set_random_state_PRNG()
```
not needed

KaiSzuttor commented on samples/widom_insertion.py:
```diff
@@ -48,7 +43,7 @@
 #############################################################
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.set_random_state_PRNG()
```
not needed

KaiSzuttor commented on samples/drude_bmimpf6.py:
```diff
@@ -75,7 +72,7 @@
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.virtual_sites = VirtualSitesRelative(have_velocity=True)
 system.set_random_state_PRNG()
```
not needed

KaiSzuttor commented on samples/ekboundaries.py:
```diff
@@ -15,19 +15,18 @@
 
-system = System(box_l=[10, 10, 10])
+system = espressomd.System(box_l=[10, 10, 10])
 system.set_random_state_PRNG()
```
not needed

KaiSzuttor commented on samples/grand_canonical.py:
```diff
@@ -55,7 +58,7 @@
 #############################################################
-system = espressomd.System(box_l=[box_l, box_l, box_l])
+system = espressomd.System(box_l=3 * [box_l])
 system.set_random_state_PRNG()
```
probably not needed

KaiSzuttor commented on samples/lj-demo.py:
```diff
@@ -88,7 +94,7 @@
 #############################################################
-system = espressomd.System(box_l=[1.0, 1.0, 1.0])
+system = espressomd.System(3 * [box_l])
 system.set_random_state_PRNG()
```
not needed

jngrad committed:
Formatting

jngrad committed:
Cleanup samples

More cleanup and apply code review feedback: remove max_num_cells,
revert espressomd.System and box_l changes, use str.format().

jngrad committed:
Minimize energy with steepest descent

Replace all occurrences of custom force_cap-based warmup code by
the steepest descent algorithm. Initialize the thermostat after
steepest descent. Don't print LJ interaction parameters again
after minimization.

jngrad committed:
Remove Tcl-specific code

Remove code writing out (sometimes incomplete) Tcl files. Write out
observables in tab-separated values files suitable for np.loadtxt()
and xmgrace plotting. Remove clutter from print() statements.

KaiSzuttor commented on doc/sphinx/io.rst:
```diff
@@ -158,35 +159,39 @@ Currently available:
-Further the files can be inspected with the GUI tool hdfview.
+Furthermore, the files can be inspected with the GUI tool hdfview.
+
+For other examples, see :file:`/samples/5hmd.py`
```
typo?

jngrad replied:
yes

KaiSzuttor commented on samples/electrophoresis.py:
```diff
@@ -17,30 +17,26 @@
 try:
     import cPickle as pickle
 except ImportError:
     import pickle
```
I think we should remove pickle here and also the lattice Boltzmann stuff. Either we show a sample with hydrodynamics and explain it or we don't. Checkpointing is also not part of this sample.

jngrad replied:
agreed, let's remove pickle and lb

jngrad committed:
Fix typos

jngrad committed:
Remove legacy code and instructions

jngrad committed:
Fix duplicate explicit target name

End links with a double underscore to make anonymous references and
avoid overshadowing explicit references (here the title reference).

pkreissl commented on samples/drude_bmimpf6.py:
```diff
@@ -16,16 +16,13 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
 Particle polarization with cold Drude oscillators on a coarse grained
```
Being pedantic, "coarse-grained simulation", I guess..

pkreissl commented on samples/electrophoresis.py:
```diff
@@ -139,41 +124,38 @@
 #############################################################
 #
-# Activate electrostatic with checkpoint example
+# Activate electrostatic
```
electrostatic**s**

pkreissl commented on samples/grand_canonical.py:
```diff
@@ -17,16 +17,19 @@
 """
-This example script performs a grand canonical simulation of a system in contact
+Perform a grand canonical simulation of a system in contact
 with a salt reservoir and ensures constant chemical potential.
```
[...] and ensure~s~ [...] / while maintaining a constant / or similar

pkreissl commented on samples/lj_liquid.py:
```diff
@@ -17,59 +17,58 @@
+* warm up and integrate.
+* write parameters, configurations and observables to files
+
+The particles in the system are of two types: Type 0 and Type 1.
```
**t**ype 0 and **t**ype 1

pkreissl commented on samples/lj_liquid.py:
```diff
@@ -17,59 +17,58 @@
+The particles in the system are of two types: Type 0 and Type 1.
+Type 0 particles interact with each other via a repulsive WCA
+interaction. Type 1 particles neither interact with themselves
+nor with Type 0 particles.
```
[...] nor with **t**ype 0 [...]

pkreissl commented on samples/lj_liquid_distribution.py:
```diff
@@ -17,44 +17,39 @@
-particles of Type 0 and Type 1 is recorded.
+Set up a Lennard-Jones fluid maintained at a fixed temperature by a
+Langevin thermostat. The particles in the system are of two types:
+Type 0 and Type 1. Type 0 particles interact with each other via a
```
[...] **t**ype 0 and **t**ype 1.

pkreissl commented on samples/lj_liquid_distribution.py:
```diff
@@ -17,44 +17,39 @@
+Langevin thermostat. The particles in the system are of two types:
+Type 0 and Type 1. Type 0 particles interact with each other via a
+repulsive WCA interaction. Type 1 particles neither interact with
+themselves nor with Type 0 particles. The distribution of minimum
```
**t**ype 0

pkreissl commented on samples/lj_liquid_distribution.py:
```diff
@@ -17,44 +17,39 @@
+Type 0 and Type 1. Type 0 particles interact with each other via a
+repulsive WCA interaction. Type 1 particles neither interact with
+themselves nor with Type 0 particles. The distribution of minimum
+distances between particles of Type 0 and Type 1 is recorded with
```
**t**ype 0 and **t**ype 1

pkreissl commented on samples/lj_liquid_structurefactor.py:
```diff
@@ -16,37 +16,40 @@
+"""
+Set up a Lennard-Jones fluid maintained at a fixed temperature by a
+Langevin thermostat. The particles in the system are of two types:
+Type 0 and Type 1. Type 0 particles interact with each other via a
```
same as above

pkreissl commented on samples/lj_liquid_structurefactor.py:
```diff
@@ -16,37 +16,40 @@
+Type 0 and Type 1. Type 0 particles interact with each other via a
+repulsive WCA interaction. Type 1 particles neither interact with
+themselves nor with Type 0 particles. The spherically averaged
+structure factor of particles of Type 0 and Type 1 is calculated
```
again

pkreissl commented on samples/load_checkpoint.py:
```diff
@@ -15,7 +15,13 @@
+* non-bonded interactions
+* particles
+* P3M parameters
+* thermostat
```
Missing the full stops at the end of each item.

pkreissl commented on samples/object_in_fluid/motivation.py:
```diff
@@ -15,35 +15,39 @@
 """
-This sample loads model red blood cells and simulates its motion
-in a complex geometry.
+Simulate the motion of flexible red blood cells in a lattice-Boltzmann fluid
```
Actually this is exactly how I would write "lattice-Boltzmann", i.e. with a lower-case "l". However everywhere else in the samples, an upper-case "L" is used. I would very much prefer consistency...

pkreissl commented on samples/reaction_ensemble.py:
```diff
@@ -17,21 +17,22 @@
+providing --reaction_ensemble or --constant_pH_ensemble as command line
+argument to the script. Be aware that in the case of the reaction ensemble,
+the dissociation constant gamma is not the thermodynamic reaction constant K,
+but rather K*1mol/l and therefore carries a unit! In the case of the of the
```
Can be left as is, would prefer "K * 1 mol/l". Your call.

pkreissl commented on samples/save_checkpoint.py:
```diff
@@ -15,7 +15,13 @@
+* non-bonded interactions
+* particles
+* P3M parameters
+* thermostat
```
Missing full stops.

pkreissl commented on samples/visualization_interactive.py:
```diff
@@ -15,15 +15,13 @@
-timed callback.
+Visualize a simulation box where the particles can be repositioned via the
+mouse and timed callbacks, and the temperature of the thermostat changed
+via the keyboard.
```
Maybe another [...] of the thermostat "can be" changed [...] for immediate clarity. Your call.

pkreissl commented on samples/visualization_lbboundaries.py:
```diff
@@ -15,7 +15,7 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-Visualization sample for Lattice Boltzmann boundary nodes.
+Visualize lattice-Boltzmann boundary nodes.
```
See comment on upper-/lower-case "l" above.

pkreissl commented on samples/visualization_poiseuille.py:
```diff
@@ -15,12 +15,12 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """
-Visualization sample for Poiseuille flow with Lattice Boltzmann.
+Visualize the Poiseuille flow in a lattice-Boltzmann fluid with an
```
Again, I fully approve of "lattice", but please be consistent...

pkreissl reviewed (changes requested):
Content-wise nice clean-up. Lgtm, now that the points raised by @KaiSzuttor have been addressed.
I only found a few more or (being a bit pedantic) less important ortho/typographical issues.

jngrad committed:
Typos

RudolfWeeber committed:
Merge branch 'python' into fix-2089

jngrad committed:
More typos

RudolfWeeber commented:
bors r+
